### PR TITLE
[EPMEDU-968]: Open the feeding point details from the Favorites page

### PR DESCRIPTION
**Pull request by ArtemKleinschmidt**
Open the feeding point details from the Favorites page

Moved HomeBottomSheet to foundation
Moved FeedingPointSheetContent and dependencies to shared
Added feeding details dialog to the favourites screen without modifications
Added "go to map" link
Extracted alpha calculation logic to make it shared

[device-2022-11-10-144435.webm](https://user-images.githubusercontent.com/32295746/201096418-2d5fd65b-a8dc-4631-8d19-d411b653c3c7.webm)

### CODE REVIEW

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt [hunk 1]
```diff
@@ -1,31 +1,72 @@
...
+        state,
+        contentAlpha,
+        onEvent,
+        buttonAlpha
+    )
+}
+
+@Composable
+private fun ScreenScaffold(
+    bottomSheetState: HomeBottomSheetState,
```
Comment (carlossantiagomenendez) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt]: Hi! The bottom sheet looks sharp, the only issue I see is that is displaying some bottom padding making it different from mapBottomSheet so please remove the '.bottomBarPadding()'. And maybe put contentAlpha and buttonAlpha in order?
Nice job!

File: shared/feature/feeding/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+dependencies {
+    implementation(projects.library.common)
+    implementation(projects.library.extensions)
+    implementation(projects.library.foundation)
+    implementation(projects.library.resources)
+    implementation(projects.library.geolocation)
+
+    implementation(libs.androidx.datastore)
```
Comment (carlossantiagomenendez) [inline shared/feature/feeding/build.gradle.kts]: I think there is no need for datastore, material.dialog.core or kapt plugin.
Please delete them.

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt [hunk 3]
```diff
@@ -13,27 +13,29 @@ import androidx.compose.ui.draw.clip
...
 import com.epmedu.animeal.resources.R
 import com.mapbox.geojson.Point
 
 @Composable
-internal fun FeedingPointSheetContent(
-    feedingPoint: FeedingPointUi,
+fun FeedingPointSheetContent(
+    feedingPoint: FeedingPointModel,
     contentAlpha: Float,
+    showOnMap: Boolean = false,
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt]: Can you rename it to isShowOnMapVisible?

File: shared/feature/feeding/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest>
+</manifest>
```
Comment (Fileek) [inline shared/feature/feeding/src/main/AndroidManifest.xml]: You can collapse it to `<manifest/>`

**Review by Fileek — CHANGES_REQUESTED** [review]
Please resolve conflicts with the develop branch

**Review by Fileek — CHANGES_REQUESTED** [review]
Please check the designs. There is no collapsed state and bottom sheet should cover whole screen

**Comment by ArtemKleinschmidt** [comment]
> 

Hey @Fileek  I've updated the dialog to be fullscreen on the favorites page. Please check the code and the attached video 
[device-2022-11-21-195113.webm](https://user-images.githubusercontent.com/32295746/203125750-5744fea9-6004-400d-9c49-7d07aab4df6c.webm)

**Comment by Fileek** [comment]
> > 
> 
> Hey @Fileek I've updated the dialog to be fullscreen on the favorites page. Please check the code and the attached video [device-2022-11-21-195113.webm](https://user-images.githubusercontent.com/32295746/203125750-5744fea9-6004-400d-9c49-7d07aab4df6c.webm)

Good job. But can I ask you to hide the bottom bar when the bottom sheet animation starts and not when it ends?

**Comment by ArtemKleinschmidt** [comment]
> > > 
> > 
> > 
> > Hey @Fileek I've updated the dialog to be fullscreen on the favorites page. Please check the code and the attached video [device-2022-11-21-195113.webm](https://user-images.githubusercontent.com/32295746/203125750-5744fea9-6004-400d-9c49-7d07aab4df6c.webm)
> 
> Good job. But can I ask you to hide the bottom bar when the bottom sheet animation starts and not when it ends?

Updated the animation 
[device-2022-11-24-142557.webm](https://user-images.githubusercontent.com/32295746/203785363-b4f56777-6006-43fb-a98f-f94e1e65d526.webm)

File: library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/FeedingPointActionButton.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package com.epmedu.animeal.foundation.dialog.bottomsheet
```
Comment (Fileek) [inline library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/FeedingPointActionButton.kt]: It is a part of the feeding feature, it shouldn't be in the library. Please move it to `feeding.presentation.ui`

File: library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt [hunk 6]
```diff
@@ -212,4 +228,33 @@ private fun Modifier.bottomSheetSwipeable(
     }
 
     return this.then(modifier)
+}
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+fun HomeBottomSheetState.contentAlphaButtonAlpha(): Pair<Float, Float> {
```
Comment (Fileek) [inline library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt]: Could you replace `Home` word with `Animeal` or remove it in the composables which you've extracted?

File: library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt [hunk 7]
```diff
@@ -118,6 +119,7 @@ fun HomeBottomSheetLayout(
     sheetElevation: Dp = ModalBottomSheetDefaults.Elevation,
     sheetBackgroundColor: Color = MaterialTheme.colors.surface,
     sheetContentColor: Color = contentColorFor(sheetBackgroundColor),
+    isShownStateEnabled: Boolean = true,
```
Comment (Fileek) [inline library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt]: Could you rename it to something like `isAlwaysExpanded` or something similar to make this boolean more clear?

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreen.kt [hunk 8]
```diff
@@ -1,14 +1,42 @@
 package com.epmedu.animeal.favourites
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreen.kt]: Please move it to `presentation` package

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenEvent.kt [hunk 9]
```diff
@@ -2,5 +2,8 @@ package com.epmedu.animeal.favourites
 
 internal sealed class FavouritesScreenEvent(open val id: Int = -1) {
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenEvent.kt]: Please move it to `presentation` package

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt [hunk 10]
```diff
@@ -1,50 +1,134 @@
 package com.epmedu.animeal.favourites
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt]: Same as above

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesState.kt [hunk 11]
```diff
@@ -1,5 +1,9 @@
 package com.epmedu.animeal.favourites
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesState.kt]: Please move it to `presentation.viewmodel` package

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesViewModel.kt [hunk 12]
```diff
@@ -13,7 +15,8 @@ import javax.inject.Inject
 internal class FavouritesViewModel @Inject constructor(
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesViewModel.kt]: Please move it to `presentation.viewmodel` package

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/model/FeedStatus.kt [hunk 13]
```diff
@@ -0,0 +1,10 @@
+package com.epmedu.animeal.feeding.presentation.model
+
+import com.epmedu.animeal.feeding.data.model.enum.AnimalState
+import com.epmedu.animeal.foundation.common.FeedStatus
+
+fun AnimalState.toFeedStatus(): FeedStatus =
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/model/FeedStatus.kt:19]: Could you please move `FeedStatus` enum to here as well and `FeedStatus` composable to the ui package? Also remove one of the `FeedStatus` composables, one is in the library and second one is in `FeedingPointSheet`

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesViewModel.kt [hunk 14]
```diff
@@ -25,11 +28,18 @@ internal class FavouritesViewModel @Inject constructor(
...
+                updateState { copy(showingFeedSpot = favourites.first { it.id == event.id }) }
+            }
+            FavouritesScreenEvent.DismissWillFeedDialog -> {
+                updateState { copy(showingWillFeedDialog = false) }
+            }
+            is FavouritesScreenEvent.ShowWillFeedDialog -> {
+                updateState { copy(showingWillFeedDialog = true) }
             }
         }
+        viewModelScope.launch { sendEvent(event) }
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesViewModel.kt]: Screen events are intended to be sent by screen. You should handle them here and react correspondingly, but you can't just send them back. Create a new sealed class for events from viewmodel - `FavouritesEvent` or `FavouritesViewModelEvent` and put it into `presentation.viewmodel` package

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenEvent.kt [hunk 15]
```diff
@@ -2,5 +2,8 @@ package com.epmedu.animeal.favourites
...
     data class FeedSpotSelected(override val id: Int) : FavouritesScreenEvent(id)
-    data class FeedSpotRemove(override val id: Int) : FavouritesScreenEvent(id)
+    data class FeedSpotChanged(override val id: Int, val isFavorite: Boolean) :
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenEvent.kt]: Why did you rename this event and add a boolean if you use it the same way? I guess we have to ask BAs if we can add feeding points back to favorites using bottom sheet

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt [hunk 16]
```diff
@@ -102,23 +201,34 @@ private fun FavouritesScreenPreview() {
...
+                        animalStatus = AnimalState.RED,
+                        animalType = AnimalType.Dogs,
+                        description = "Hungry dog",
+                        lastFeeder = Feeder(id = 0, "Fred", "12:00"),
+                        location = MapLocation.Tbilisi,
+                    ),
                 )
-            )
+            ),
+            rememberHomeBottomSheetState(HomeBottomSheetValue.Hidden)
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt]: Don't use remember in preview, it's static anyway

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt [hunk 17]
```diff
@@ -102,23 +201,34 @@ private fun FavouritesScreenPreview() {
...
 @AnimealPreview
 @Composable
 private fun FavouritesScreenEmptyPreview() {
     AnimealTheme {
         FavouritesScreenUI(
             FavouritesState(
                 emptyList()
-            )
+            ),
+            rememberHomeBottomSheetState(HomeBottomSheetValue.Hidden)
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt]: Same as above

File: library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt [hunk 18]
```diff
@@ -1,6 +1,7 @@
-package com.epmedu.animeal.home.presentation.ui
+package com.epmedu.animeal.foundation.dialog.bottomsheet
```
Comment (Fileek) [inline library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt]: It is a part of the feeding feature, it shouldn't be in the library. Please move it to feeding.presentation.ui

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreen.kt [hunk 19]
```diff
@@ -1,14 +1,42 @@
...
+    FavouritesScreenUI(state, bottomSheetState, onEvent = {
+        viewModel.handleEvents(it)
+    })
 
-    FavouritesScreenUI(state, onEvent = { viewModel.handleEvents(it) })
+    LaunchedEffect(Unit) {
+        viewModel.events.collect {
+            when (it) {
+                is FavouritesScreenEvent.FeedSpotSelected -> {
+                    launch {
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreen.kt]: Why do you launch child coroutine here, but not below?

**Review by Fileek — CHANGES_REQUESTED** [review]

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt [hunk 20]
```diff
@@ -13,27 +13,30 @@ import androidx.compose.ui.draw.clip
...
 import com.epmedu.animeal.resources.R
 import com.mapbox.geojson.Point
 
 @Composable
-internal fun FeedingPointSheetContent(
-    feedingPoint: FeedingPointUi,
+fun FeedingPointSheetContent(
+    feedingPoint: FeedingPointModel,
     contentAlpha: Float,
+    expandToFullScreen: Boolean = false,
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt]: You should use modifier and pass it to the column instead of Spacer depended on this boolean

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt (hunk 16 above)
Comment (ArtemKleinschmidt) [reply feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt]: @Fileek  State is an obligatory parameter, need to pass something here

File: library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt (hunk 18 above)
Comment (ArtemKleinschmidt) [reply library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt]: @Fileek  it is not a part of feeding feature. It is a custom bottom sheet dialog, that can be used with custom content anywhere in the app

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedStatusItem.kt [hunk 21]
```diff
@@ -48,11 +44,18 @@ fun FeedStatus(
...
 
-enum class FeedStatus(
-    @StringRes val titleId: Int,
-    @DrawableRes val iconId: Int,
-    val color: Color
-) {
-    GREEN(R.string.feed_status_green, R.drawable.ic_face_happy, CustomColor.StatusGreen),
-    RED(R.string.feed_status_red, R.drawable.ic_face_upset, CustomColor.StatusRed);
+@Composable
+@Preview
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedStatusItem.kt]: Please use AnimealPreview

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedStatusItem.kt [hunk 22]
```diff
@@ -48,11 +44,18 @@ fun FeedStatus(
...
+fun FeedStatusItemGreenPreview() {
+    FeedStatusItem(
+        status = FeedStatus.GREEN
+    )
+}
+
+@Composable
+@Preview
+fun FeedStatusItemRedPreview() {
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedStatusItem.kt]: It's better to put different previews of one small ui component in one preview using column and dividers

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt (hunk 16 above)
Comment (Fileek) [reply feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt]: Just create it as an instance of the class

File: shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt [hunk 23]
```diff
@@ -18,38 +18,42 @@ import androidx.compose.ui.text.style.TextOverflow
...
-            .wrapContentHeight()
             .background(MaterialTheme.colors.surface)
             .padding(
                 top = 8.dp,
                 bottom = 110.dp,
                 start = 24.dp,
                 end = 24.dp,
-            ),
+            )
+            .then(modifier),
```
Comment (Fileek) [inline shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt]: You should use this modifier in the beginning as the receiver of the chain

File: library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/AnimealBottomSheet.kt [hunk 24]
```diff
@@ -1,6 +1,7 @@
-package com.epmedu.animeal.home.presentation.ui
+package com.epmedu.animeal.foundation.dialog.bottomsheet
```
Comment (Fileek) [inline library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/AnimealBottomSheet.kt]: Shouldn't be in `dialog`

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesState.kt [hunk 25]
```diff
@@ -0,0 +1,10 @@
+package com.epmedu.animeal.favourites.presentation.viewmodel
+
+import com.epmedu.animeal.feeding.domain.model.FeedingPoint
+
+data class FavouritesState(
+    val favouritesSnapshot: List<FeedingPoint> = emptyList(),
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesState.kt]: Looks like you don't need this list on UI. Move it to the viewmodel as a private variable instead

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesViewModel.kt [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+        return if (state.showingFeedSpot?.id == event.id) {
+            state.showingFeedSpot.copy(isFavourite = event.isFavorite)
+        } else {
+            state.showingFeedSpot
+        }
+    }
+
+    private fun updateSnapshot(event: FavouritesScreenEvent.FeedSpotChanged): MutableList<FeedingPoint> {
+        val snapshot = mutableListOf<FeedingPoint>()
+        state.favouritesSnapshot.forEach { feedingPoint ->
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesViewModel.kt]: Can be simplified with map operator

File: feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesViewModel.kt [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+            snapshot += if (event.id == feedingPoint.id) {
+                feedingPoint.copy(isFavourite = event.isFavorite)
+            } else {
+                feedingPoint.copy()
```
Comment (Fileek) [inline feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesViewModel.kt]: Redundant `copy()`

**Review by Fileek — APPROVED** [review]

**Review by VEZE — APPROVED** [review]
